include/Sample.h

```diff
@@ -0,0 +1,158 @@
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef LMMS_SAMPLE_H
+#define LMMS_SAMPLE_H
+
+#include <cmath>
+#include <memory>
+
+#include "Note.h"
+#include "SampleBuffer.h"
+#include "lmms_export.h"
+
+#ifdef __MINGW32__
+#include <mingw.mutex.h>
+#include <mingw.shared_mutex.h>
+#else
+#include <mutex>
+#include <shared_mutex>
+#endif
+
+class QPainter;
+class QRect;
+
+namespace lmms {
+class LMMS_EXPORT Sample
+{
+public:
+	// values for buffer margins, used for various libsamplerate interpolation modes
+	// the array positions correspond to the converter_type parameter values in libsamplerate
+	// if there appears problems with playback on some interpolation mode, then the value for that mode
+	// may need to be higher - conversely, to optimize, some may work with lower values
+	static constexpr auto s_interpolationMargins = std::array<int, 5>{64, 64, 64, 4, 4};
+
+	enum class Loop
+	{
+		Off,
+		On,
+		PingPong
+	};
+
+	class LMMS_EXPORT PlaybackState
+	{
+	public:
+		PlaybackState(bool varyingPitch = false, int interpolationMode = SRC_LINEAR)
+			: m_varyingPitch(varyingPitch)
+			, m_interpolationMode(interpolationMode)
+		{
+			auto error = 0;
+			m_resampleState = src_new(m_interpolationMode, DEFAULT_CHANNELS, &error);
+			if (error != 0)
+			{
+				throw std::runtime_error{"Error creating resample state: " + std::string{src_strerror(error)}};
+			}
+		}
+
+		~PlaybackState() noexcept { src_delete(m_resampleState); }
+
+		auto frameIndex() const -> f_cnt_t { return m_frameIndex; }
+		auto varyingPitch() const -> bool { return m_varyingPitch; }
+		auto backwards() const -> bool { return m_backwards; }
+		auto interpolationMode() const -> int { return m_interpolationMode; }
+
+		auto setFrameIndex(f_cnt_t frameIndex) -> void { m_frameIndex = frameIndex; }
+		auto setVaryingPitch(bool varyingPitch) -> void { m_varyingPitch = varyingPitch; }
+		auto setBackwards(bool backwards) -> void { m_backwards = backwards; }
+
+	private:
+		f_cnt_t m_frameIndex = 0;
+		bool m_varyingPitch = false;
+		bool m_backwards = false;
+		SRC_STATE* m_resampleState = nullptr;
+		int m_interpolationMode = SRC_LINEAR;
+		friend class Sample;
+	};
+
+	Sample() = default;
+	Sample(const QByteArray& base64, int sampleRate = Engine::audioEngine()->processingSampleRate());
+	Sample(const sampleFrame* data, int numFrames, int sampleRate = Engine::audioEngine()->processingSampleRate());
+	Sample(const Sample& other);
+	Sample(Sample&& other);
+	explicit Sample(const QString& audioFile);
+	explicit Sample(std::shared_ptr<const SampleBuffer> buffer);
+
+	auto operator=(const Sample&) -> Sample&;
+	auto operator=(Sample&&) -> Sample&;
+
+	auto play(sampleFrame* dst, PlaybackState* state, int numFrames, float desiredFrequency = DefaultBaseFreq,
+		Loop loopMode = Loop::Off) const -> bool;
+	auto visualize(QPainter& p, const QRect& dr, int fromFrame = 0, int toFrame = 0) const -> void;
```

Comment:
> I realise it was like this before, but this doesn't belong in the core.

Was already aware of this, but left it since I had ideas to create a new widget that resembles a sample waveform, and not just have some `visualize` function. However, creating that new widget I feel is a bit much for this PR, and would cause more changes that I rather not deal with here.

I will still move `visualize` out of `Sample` here though, but I think in the future it could make more sense to make this a widget.